src/Avalonia.X11/X11WindowModes/XEmbedClientWindowMode.cs

```diff
@@ -0,0 +1,179 @@
+                    ptr2 = new ((int)message),
+                    ptr3 = detail,
+                    ptr4 = data1,
+                    ptr5 = data2
+                }
+            };
+            XSendEvent(X11.Display, _currentEmbedder, false,
+                new IntPtr((int)(EventMask.NoEventMask)), ref xev);
+        }
+        
+        static XEmbedClientWindowMode()
+        {
+            KeyboardDevice.Instance.PropertyChanged += (_, args) =>
+            {
+                if (args.PropertyName == nameof(KeyboardDevice.Instance.FocusedElement))
+                {
+                    if (KeyboardDevice.Instance.FocusedElement is Visual visual
+                        && visual.VisualRoot is EmbeddableControlRoot root
+                        && root.PlatformImpl is X11Window window
+                        && window._mode is XEmbedClientWindowMode xembedMode
+                        && xembedMode._currentEmbedder != IntPtr.Zero)
+                    {
+                        xembedMode._savedFocus = KeyboardDevice.Instance.FocusedElement;
+                        xembedMode.SendXEmbedMessage(XEmbedMessage.RequestFocus);
+                    }
+                }
+            };
+        }
+
+        void Reset()
+        {
+            _embedderActivated = false;
+            _focusedInEmbedder = false;
+            _disabled = false;
+            UpdateActivation();
+        }
+        
+        void OnXEmbedMessage(IntPtr time, XEmbedMessage message, IntPtr detail, IntPtr data1, IntPtr data2)
+        {
+            if (message == XEmbedMessage.EmbeddedNotify)
+            {
+                Reset();
+                _currentEmbedder = data1;
+            }
+            else if (message == XEmbedMessage.FocusIn)
+            {
+                _focusedInEmbedder = true;
+                UpdateActivation();
+            }
+            else if (message == XEmbedMessage.FocusOut)
+            {
+                _focusedInEmbedder = false;
+                UpdateActivation();
+            }
+            else if (message == XEmbedMessage.WindowActivate)
+            {
+                _embedderActivated = true;
+                UpdateActivation();
+            }
+            else if (message == XEmbedMessage.WindowDeactivate)
+            {
+                _embedderActivated = false;
+                UpdateActivation();
+            }
+            else if (message == XEmbedMessage.ModalityOn)
+                _disabled = true;
+            else if (message == XEmbedMessage.ModalityOff)
+                _disabled = false;
+        }
+
+        private void UpdateActivation()
+        {
+            var active = _focusedInEmbedder && _embedderActivated;
+
+            if (active)
+            {
+                ((FocusManager?)Root?.FocusManager)?.SetFocusScope(Root);
+                if (_savedFocus != null) 
+                    KeyboardDevice.Instance?.SetFocusedElement(_savedFocus, NavigationMethod.Unspecified, KeyModifiers.None);
```

Comment:
Any particular reason to use KeyboardDevice? I don't see why we would avoid logical focus here, especially when we set focus scope explicitly.

You can use stable API and call `_savedFocus.Focus(NavigationMethod.Unspecified, KeyModifiers.None)`. Which internally will also set focus correct scope, so you don't need to do it manually.